src/query/api/v1/middleware/source.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"go.uber.org/zap"
+
+	"github.com/m3db/m3/src/query/source"
+	"github.com/m3db/m3/src/query/util/logging"
+	"github.com/m3db/m3/src/x/headers"
+	xhttp "github.com/m3db/m3/src/x/net/http"
+)
+
+var errInvalidSourceHeader = xhttp.NewError(
+	fmt.Errorf("invalid %s header", headers.SourceHeader),
+	http.StatusBadRequest)
+
+// SourceOptions are the options for the source middleware.
+type SourceOptions struct {
+	Deserializer source.Deserializer
+}
+
+// Source adds the headers.SourceHeader value to the request context.
+// Installing this middleware function allows application code to access the typed source value using FromContext.
+// Additionally a source log field is added to the request scope logger.
+func Source(opts Options) mux.MiddlewareFunc {
```

Comment:
this had to move to the `middleware` pkg so the `SourceOptions` could be part of the `MiddlewareOptions`.

This is a tradeoff for allowing routes to overrides options for middleware. All middleware will need to be the `middleware` pkg, instead of individual pkgs. I think this is fine, since `middleware` fns will just delegate to exposed functions from other pkgs, like `source.NewContext`